File: core/src/main/java/com/jetbrains/tmp/learning/courseFormat/StepType.java
package com.jetbrains.tmp.learning.courseFormat;

import org.jetbrains.annotations.NotNull;

/**
 * @author meanmail
 */
public enum StepType {
    UNKNOWN, CODE, TEXT, VIDEO, CHOICE, STRING, SORTING, MATCHING, NUMBER, DATASET;

    @NotNull
    public static StepType of(String name) {
        try {
            return valueOf(name.toUpperCase());
        } catch (IllegalArgumentException e) {
            return UNKNOWN;
        }
    }

    @Override
    public String toString() {
        return name().toLowerCase();
    }
}

File: stepik-core/src/main/kotlin/org/stepik/core/courseFormat/StepType.kt
package org.stepik.core.courseFormat


enum class StepType(val typeName: String) {
    UNKNOWN("unknown"),
    CODE("code"),
    TEXT("text"),
    VIDEO("video"),
    CHOICE("choice"),
    STRING("string"),
    SORTING("sorting"),
    MATCHING("matching"),
    NUMBER("number"),
    DATASET("dataset"),
    TABLE("table"),
    FILL_BLANKS("fill-blanks"),
    MATH("math"),
    FREE_ANSWER("free-answer");

    override fun toString(): String {
        return typeName
    }

    companion object {

        private val map by lazy {
            values().associateBy { it.typeName }
        }

        fun of(typeName: String?) = map.getOrDefault(typeName, UNKNOWN)
    }
}
